src/plugins/dashboard/public/application/utils/create_dashboard_container.tsx

```diff
@@ -0,0 +1,518 @@
+};
+
+const handleDashboardContainerChanges = (
+  dashboardContainer: DashboardContainer,
+  appState: DashboardAppStateContainer,
+  dashboardServices: DashboardServices,
+  dashboard: Dashboard
+) => {
+  let dirty = false;
+  if (!appState) {
+    return;
+  }
+  const appStateData = appState.getState();
+  const savedDashboardPanelMap: { [key: string]: SavedDashboardPanel } = {};
+  const { opensearchDashboardsVersion } = dashboardServices;
+  const input = dashboardContainer.getInput();
+  appStateData.panels.forEach((savedDashboardPanel) => {
+    if (input.panels[savedDashboardPanel.panelIndex] !== undefined) {
+      savedDashboardPanelMap[savedDashboardPanel.panelIndex] = savedDashboardPanel;
+    } else {
+      // A panel was deleted.
+      dirty = true;
+    }
+  });
+
+  const convertedPanelStateMap: { [key: string]: SavedDashboardPanel } = {};
+  Object.values(input.panels).forEach((panelState) => {
+    if (savedDashboardPanelMap[panelState.explicitInput.id] === undefined) {
+      dirty = true;
+    }
+    convertedPanelStateMap[panelState.explicitInput.id] = convertPanelStateToSavedDashboardPanel(
+      panelState,
+      opensearchDashboardsVersion
+    );
+    if (
+      !isEqual(
+        convertedPanelStateMap[panelState.explicitInput.id],
+        savedDashboardPanelMap[panelState.explicitInput.id]
+      )
+    ) {
+      // A panel was changed
+      // Do not need to care about initial migration here because the version update
+      // is already handled in migrateAppState() when we create state container
+      dirty = true;
+    }
+  });
+
+  const newAppState: { [key: string]: any } = {};
+  if (dirty) {
+    newAppState.panels = Object.values(convertedPanelStateMap);
+    dashboard.isDirty = true;
+  }
+  if (input.isFullScreenMode !== appStateData.fullScreenMode) {
+    newAppState.fullScreenMode = input.isFullScreenMode;
+  }
+  if (input.expandedPanelId !== appStateData.expandedPanelId) {
+    newAppState.expandedPanelId = input.expandedPanelId;
+  }
+  if (!isEqual(input.query, migrateLegacyQuery(appState.get().query))) {
+    newAppState.query = input.query;
+  }
+
+  if (Object.keys(newAppState).length > 0) {
+    appState.transitions.setDashboard({ ...newAppState });
+  }
+
+  // event emit dirty?
+};
+
+export const refreshDashboardContainer = (
+  dashboardContainer: DashboardContainer,
+  dashboardServices: DashboardServices,
+  savedDashboardId: string,
+  dashboard: Dashboard,
+  appStateData?: DashboardAppState
+) => {
+  if (!appStateData) {
+    return;
+  }
+
+  const currentDashboardInput = getDashboardInputFromAppState(
+    appStateData,
+    dashboardServices,
+    savedDashboardId
+  );
+
+  const changes = getChangesForContainerStateFromAppState(
+    dashboardContainer,
+    currentDashboardInput
+  );
+
+  if (changes) {
+    dashboardContainer.updateInput(changes);
+
+    if (changes.timeRange || changes.refreshConfig) {
```

Comment:
I think line 508-512 needs to be:
```suggestion
    if (changes.timeRange || changes.refreshConfig) {
              if (dashboardInstance.timeRestore) {
                dashboard.isDirty = true;
              }
            }
```